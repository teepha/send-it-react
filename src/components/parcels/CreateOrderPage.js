import React from "react";
import Spinner from "react-md-spinner";
import { Link, withRouter } from "react-router-dom";
import { connect } from "react-redux";
import { toast } from "react-toastify";
import { createParcelOrder } from "../../actions/parcelsActions";

export class CreateOrder extends React.Component {
  state = {
    pickupLocation: "",
    destination: "",
    recipientName: "",
    recipientPhone: ""
  };

  shouldComponentUpdate(nextProps) {
    if (this.props.parcels.length !== nextProps.parcels.length) {
      if (this.props.error !== nextProps.error) {
        const errorString = nextProps.error;
        toast.warn(errorString);
      } else {
        toast.success("Parcel Order Created Successfully!");
        this.props.history.push("/user-profile");
      }
    }
    return true;
  }

  handleCreateOrder = e => {
    e.preventDefault();
    const { user } = this.props;
    const userId = user.id;
    this.props.createParcelOrder(userId, this.state);
  };

  handleInputChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    const { processing } = this.props;
    return (
      <div className="main-create-order-page">
        <div className="create-order-wrapper">
          <div className="create-order-page">
            <Link to="./user-profile">
              <i className="fas fa-arrow-left" />
            </Link>
            <h1>New Order</h1>
            <form onSubmit={this.handleCreateOrder} id="create-order-form">
              <label>Pick Up Location </label>
              <input
                required
                pattern="[a-zA-Z0-9_&-]+([ ]?[a-zA-Z0-9_&-]+)*"
                minLength="10"
                name="pickupLocation"
                type="text"
                value={this.state.pickup_location}
                onChange={this.handleInputChange}
              />
              <br />
              <label>Destination </label>
              <input
                required
                pattern="[a-zA-Z0-9_&-]+([ ]?[a-zA-Z0-9_&-]+)*"
                minLength="10"
                name="destination"
                type="text"
                value={this.state.destination}
                onChange={this.handleInputChange}
              />
              <br />
              <label>Recipient Name </label>
              <input
                required
                pattern="^[A-Za-z]+$"
                name="recipientName"
                type="text"
                value={this.state.recipient_name}
                onChange={this.handleInputChange}
              />
              <br />
              <label>Recipient Phone Number </label>
              <input
                required
                minLength="10"
                name="recipientPhone"
                type="text"
                value={this.state.recipient_phone}
                onChange={this.handleInputChange}
              />
              <br />
              <button className="button" type="submit" disabled={processing}>
                {processing ? (
                  <Spinner size={18} singleColor="#fff" />
                ) : (
                  "Create Order"
                )}
              </button>
              <h4 id="error-msg" />
            </form>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  processing: state.user.isLoading,
  user: state.user.userData,
  parcels: state.parcels.data,
  error: state.parcels.error
});

const mapDispatchToProps = {
  createParcelOrder
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(CreateOrder)
);
