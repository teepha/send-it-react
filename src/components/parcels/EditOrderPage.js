import React from "react";
import Spinner from "react-md-spinner";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { toast } from "react-toastify";
import { updateParcelOrder, getSingleParcel } from "../../actions/parcelsActions";

class EditOrder extends React.Component {
  state = {
    pickupLocation: "",
    destination: "",
    recipientName: "",
    recipientPhone: "",
  };

  componentDidMount() {
    this.props.getSingleParcel(this.props.match.params.id);
  }

  shouldComponentUpdate = (nextProps) => {
    if (!this.state.pickupLocation) {
      this.setState({
        pickupLocation: nextProps.parcel.pickup_location,
        destination: nextProps.parcel.destination,
        recipientName: nextProps.parcel.recipient_name,
        recipientPhone: nextProps.parcel.recipient_phone,
      });
    } else if (this.props.parcel !== nextProps.parcel) {
      if (nextProps.parcel) {
        toast.success("Parcel Order Updated Successfully!");
        this.props.history.push("/user-profile");
      } else {
        const errorString = nextProps.error;
        toast.warn(errorString);
      }
    }
    return true;
  };

  handleEditOrder = (e) => {
    const { id } = this.props.match.params;
    e.preventDefault();
    this.props.updateParcelOrder(id, this.state);
  };

  handleInputChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    const { processing } = this.props;
    return (
      <div className="main-edit-order-page">
        <div className="edit-order-wrapper">
          <div className="edit-order-page">
            <Link to="/user-profile">
              <i className="fas fa-arrow-left" />
            </Link>
            <h1>Edit Order</h1>

            <form onSubmit={this.handleEditOrder} id="edit-order-form">
              <label>Pick Up Location </label>
              <input
                required
                name="pickupLocation"
                type="text"
                minLength="10"
                value={this.state.pickupLocation}
                onChange={this.handleInputChange}
              />
              <br />
              <label>New Destination </label>
              <input
                required
                name="destination"
                type="text"
                minLength="10"
                value={this.state.destination}
                onChange={this.handleInputChange}
              />
              <br />
              <label>Recipient Name </label>
              <input
                required
                name="recipientName"
                type="text"
                pattern="^[A-Za-z]+$"
                value={this.state.recipientName}
                onChange={this.handleInputChange}
              />
              <br />
              <label>Recipient Phone Number </label>
              <input
                required
                name="recipientPhone"
                type="text"
                minLength="10"
                value={this.state.recipientPhone}
                onChange={this.handleInputChange}
              />
              <br />
              <button
                className="button"
                type="submit"
                disabled={processing}
              >
                {processing ? (
                  <Spinner
                    size={18}
                    singleColor="white"
                  />
                ) : "Proceed"}
              </button>
              <h4 id="error-msg" />
            </form>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ user, parcels }, ownProps) => {
  const props = parseInt(ownProps.match.params.id, 10);
  return {
    processing: user.isProcessing,
    user: user.userData,
    parcel: parcels.data.find(parcel => parcel.id === props),
    error: parcels.error,
  };
};

const mapDispatchToProps = id => ({
  updateParcelOrder,
  getSingleParcel,
});

export default connect(
  mapStateToProps,
  mapDispatchToProps(),
)(EditOrder);
