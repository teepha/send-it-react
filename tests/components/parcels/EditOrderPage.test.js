import React from "react";
import { shallow } from "enzyme";
import { EditOrder } from "../../../src/components/parcels/EditOrderPage";

describe("EditOrder Component Test", () => {
  const setUp = () => {
    const props = {
      updateParcelOrder: jest.fn(() => Promise.resolve()),
      getSingleParcel: jest.fn(() => Promise.resolve()),
      processing: false,
      user: {
        id: 1,
        role: "admin"
      },
      parcel: {},
      error: "",
      history: {
        push: jest.fn()
      },
      match: {
        params: {
          id: 3
        }
      }
    };
    const state = {
      pickupLocation: "",
      destination: "",
      recipientName: "",
      recipientPhone: ""
    };
    return {
      shallowWrapper: shallow(<EditOrder {...props} />),
      state,
      props
    };
  };

  const event = {
    target: {
      value: ""
    },
    preventDefault: jest.fn()
  };

  it("render without crashing", () => {
    const { shallowWrapper } = setUp();
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it("should display <Spinner /> component when `isProcessing` is set to true", () => {
    const { shallowWrapper, props } = setUp();
    shallowWrapper.setProps({
      ...props,
      processing: true
    });
    expect(shallowWrapper.find("MDSpinner")).toHaveLength(1);
  });

  it("invokes handleInputChange method", () => {
    const { shallowWrapper } = setUp();
    const handleInputChangeSpy = jest.spyOn(
      shallowWrapper.instance(),
      "handleInputChange"
    );
    shallowWrapper.instance().handleInputChange(event);
    expect(handleInputChangeSpy).toHaveBeenCalled();
  });

  it("invokes handleEditOrder method", () => {
    const { shallowWrapper } = setUp();
    const handleEditOrderSpy = jest.spyOn(
      shallowWrapper.instance(),
      "handleEditOrder"
    );
    shallowWrapper.instance().handleEditOrder(event);
    expect(handleEditOrderSpy).toHaveBeenCalledWith(event);
  });
});
