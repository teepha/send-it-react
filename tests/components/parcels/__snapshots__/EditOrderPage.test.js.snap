// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditOrder Component Test render without crashing 1`] = `
<div
  className="main-edit-order-page"
>
  <div
    className="edit-order-wrapper"
  >
    <div
      className="edit-order-page"
    >
      <Link
        replace={false}
        to="/user-profile"
      >
        <i
          className="fas fa-arrow-left"
        />
      </Link>
      <h1>
        Edit Order
      </h1>
      <form
        id="edit-order-form"
        onSubmit={[Function]}
      >
        <label>
          Pick Up Location 
        </label>
        <input
          minLength="10"
          name="pickupLocation"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
        <br />
        <label>
          New Destination 
        </label>
        <input
          minLength="10"
          name="destination"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
        <br />
        <label>
          Recipient Name 
        </label>
        <input
          name="recipientName"
          onChange={[Function]}
          pattern="^[A-Za-z]+$"
          required={true}
          type="text"
          value=""
        />
        <br />
        <label>
          Recipient Phone Number 
        </label>
        <input
          minLength="10"
          name="recipientPhone"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
        <br />
        <button
          className="button"
          disabled={false}
          type="submit"
        >
          Proceed
        </button>
        <h4
          id="error-msg"
        />
      </form>
    </div>
  </div>
</div>
`;
